### DISCUSSION TITLE
How React Suspense and Concurrent Mode Fit Into The Picture

### DISCUSSION THREAD

**Discussion by andria-dev** [statelyai/xstate#1746] [Ideas]
Without changing XState to suit React's needs, **what ways do we think would be possible to allow support for Suspense to be used with things such as invoked promises**? I do believe this will benefit more than just React users in the future as UI as a first-class feature potentially becomes a more popular idea.

## Idea

I'd imagine that extremely basic support would involve exposing the promise that was invoked so that it could be thrown, for example:

```jsx
async function getTheThing(ctx) {
  // Do some asynchronous stuff, maybe a fetch or two
}

const promiseMachine = Machine({
  /* ... */,
  states: {
    loading: {
      invoke: {
        src: getTheThing,
        onDone: /* save the data */,
        onError: /* save the error */,
      },
    },
    /* More states... */
})

function MyComponent() {
  const [current] = useMachine(promiseMachine)

  // MAGIC HERE. There currently isn't a `.invoked.src` or even a `.invoked`
  if (current.matches('loading'))
    throw current.invoked.src

  // maybe throw an error if one occurred
  // then if everything is good and done, render stuff
  return <p>{current.context.theThing}</p>
}

// Somewhere else:
<Suspense fallback={<Spinner label="Getting the thing, hold on..." />}>
  <MyComponent />
</Suspense>
```

To re-iterate, I don't think XState needs to really change hardly anything, just being able to see the promise should be enough. As of right now, I can't find any way to view the promise from the result of `useMachine()` at all — I'm assuming `useMachine` isn't the problem here.

## What do we have currently?

I've found that you can see some info about your invoked thing by looking at `.activities`, sadly it doesn't provide what we need. Here is an example of what can be found in `.activites`:

```json5
// current.activities ===
{
  "image-loading-machine.fetching:invocation[0]": {
    type: "xstate.start",
    activity: {
      id: "image-loading-machine.fetching:invocation[0]",
      src: "image-loading-machine.fetching:invocation[0]",
      type: "xstate.invoke",
    },
  },
}
```

If you'll notice above, there is a `.src` on the activity itself but it is just the `id` again for some reason.

## Discussion

If anyone has any other ideas on how this could work feel free to bring your ideas, this discussion is open.

**Comment by Andarist** [comment]
I'm not sure if `useMachine` is the best companion to Suspense. Or at least not in a way that you have demoed it.

`useMachine` keeps state **local** to a component but, from what I understand, when you suspend you throw away your local state. So to make your pseudo-code work with Suspense you would have to externalize your machine state anyway - or persist it and rehydrate with `useMachine`. 

Quite frankly the Suspense, even if partially working already, is not mainstream yet and patterns etc are not yet fully emerged. I'm not sure if we are best suited here for figuring them out - I would much more prefer to just learn from others' findings and incorporate best ideas to XState. That being said - I'm thinking about this from time to time, there is just so many things that we still can do for the XState itself that are not as uncertain as integrating with Suspense that this is pretty low on my priority list.

**Reply by andria-dev** [reply]
I hadn't thought of that, thank you! So for that part, that would be more specific to the @xstate/react implementation I believe. The @xstate/react implementation just needs the base XState library to allow it access to a promise that will resolve or reject when the invoked thing finishes — it would be easiest to start with promises then perhaps in the future, if it makes sense, discuss support for the other invokable types.

As I've seen it work in React — I've only recently started working with Suspense — as long as you create the resource one level above where you suspend from, React seems to handle the rest flawlessly. How does this potential solution look:

```jsx
function Outer() {
  const [current] = useMachine(promiseMachine)

  // resource could hold the promise
  // context is passed for when it's done
  return (
    <Inner resource={current?.invoked?.src} context={current.context} />
  )
}

function Inner({resource, context}) {
  if (resource !== null) throw resource

  // use context here
  // to render something
}

// Somewhere else
<Suspense fallback={/* Spinner */}>
  <Outer />
</Suspense>
```

I believe this should work fine, however, if having the `<Suspense>` outside of `<Outer>` is a problem, for some cases, it can just be moved inside of it.

**Reply by andria-dev** [reply]
And I understand your hesitation to work on adding support for Suspense when it isn't even out yet. However, I feel that having this discussion could be beneficial, especially if it could potentially benefit other use cases outside of Suspense.

**Reply by Andarist** [reply]
Could u prepare a runnable demo (codesandbox preferably) of the code u have posted? I could play around quickly with it then to check out couple of things

EDIT: just with a regular promise, not the useMachine

**Reply by andria-dev** [reply]
@Andarist I just made [the CodeSandbox you requested](https://codesandbox.io/s/competent-https-r4szo?file=/src/App.js). It's a little messy but I think it gets the point across with semi-decent logic implemented.

If the fetch request is too fast on your connection you can always visit the Network tab and emulate a "Slow 3G" connection.

**Reply by Andarist** [reply]
Thank you for the sandbox - as far as I can tell I was wrong and the state of the suspended tree is not thrown away which is a good thing for us 👍 

I'm wondering though if exposing promises that could be thrown would be a good thing. This only works with promises so we'd have to provide means of promisifying other data types anyway (machines, observables, etc). What if you could promisify "conditions"? Like - await a particular state or await smth being added to context?

**Reply by davidkpiano** [reply]
> What if you could promisify "conditions"? Like - await a particular state or await smth being added to context?

I've thought about that too... the tricky part is that React has no concept of "multi-async", i.e., the idea that multiple values can be received asynchronously instead of just one (basically observables). State machines are closer to observables than promises. But we can make a promise that awaits some condition:

```js
const waitForCondition = (condition) => new Promise(res => {
  service.subscribe(state => {
    if (condition(state)) {
      res(state);
    }
  });

  // possibly do a timeout and/or reject if state machine stops without reaching condition
});
```

**Reply by andria-dev** [reply]
@Andarist I've already devised a way to promisify subscriptions. I've also found a workaround for the idea of "multi-async" in React with subscriptions @davidkpiano 

### Promisify subscription

```javascript
function readResource({status, suspender, result}) {
	switch (status) {
		case 'loading':
			throw suspender
		case 'error':
			throw result
		case 'success':
		default:
			return result
	}
}

function createResourceFromSubscription(subscribe) {
	let status = 'loading'
	let result

	const suspender = new Promise((resolve, reject) => {
		const unsubscribe = subscribe((data, error) => {
			unsubscribe()
			if (error) {
				status = 'error'
				result = error
				reject(error)
			} else {
				status = 'success'
				result = data
				resolve(data)
			}
		})
	})

	return {
		read() {
			return readResource({status, suspender, result})
		},
	}
}

// I made this for Firebase Authentication:
// const resource = createResourceFromSubscription(auth.onAuthStateChanged.bind(auth))
```

Now, this has created a single promise from the subscription that will only fire once. To get the newer values we have to use a `useEffect()`. Specifically, I've been working with Firebase's Firestore API and I devised the following hook for it.

### Subscribe to more values after the resource's value

```jsx
export function useDocumentResource(documentRef, resource) {
	const [result, setResult] = useState(resource.read())

	useEffect(() => {
		let isFirst = true
		// returns the "unsubscribe" function
		return documentRef.onSnapshot((snapshot, error) => {
			if (isFirst) {
				isFirst = false
				return
			}

			if (error) console.warn(error)
			else setResult(snapshot.data())
		})
	}, [documentRef])

	return result
}
```

Now, I wrote this `useDocumentResource()` hook for use with my own `createDocumentResource()` function that uses `documentRef.get()` promise so the DX could be improved.

To use the `useDocumentResource()` hook, I would give it the resource I'd create a resource 1 level above and pass it down through props. Then, I'd use this hook to consume the resource and create the `useEffect()` subscription that would continue to update the returned value.

**Reply by Andarist** [reply]
@davidkpiano the given piece of code is similar in terms of conditions and promisifying to what I had in mind. To accommodate the @ChrisBrownie55 use case the usage could look like this:
```js
function MyComponent() {
  const [current] = useMachine(promiseMachine)

  if (current.matches('loading'))
    throw createSuspendableCondition(state => state.context.theThing)

  return <p>{current.context.theThing}</p>
}
```

**Note:** this would only be useful for already committed machines/components - I've done additional tests and if we throw during initial render then the state is lost therefore my previous comment about the need for externalizing machine state etc still stands. 

As to errors - I would have to re-check it but I'm pretty sure that throwing those would actually lose the local state so I'm not sure in what situations that would be desirable (unless you really want to handle a scenario using a React's error boundary).

**Reply by andria-dev** [reply]
@Andarist I think errors would just be up to the developer still as they can still do `onError` with their invokable and then place the error into context and/or go to an error state.

And I don't think you need to externalize the machine's state still unless what you mean by that is to move it up one level. I've yet to find a use case where the following pattern doesn't work:

```jsx
function Outer() {
  // do stateful stuff like useMachine
  // create the resource
  return <Inner resource={theResource} />
}

function Inner({resource}) {
  // I can now use the resource without any issues of losing state
  // This means I can throw the promise or I can get a value and render whatever
}
```

For an XState machine, the pattern might look more like what you're doing but still in two components:

```jsx
function Outer() {
  const [current, send] = useMachine(promiseMachine)
  return <Inner current={current} send={send} suspender={createSuspendableCondition(/* ... */)} />
}

function Inner({current, send, suspender}) {
  if (current.matches('loading'))
    throw suspender

  return /* ... */
}
```

Note that I had to move the suspender, the promise, up to `<Outer>` so that it wouldn't be blown away when `<Inner>` suspends.

**Reply by Andarist** [reply]
> And I don't think you need to externalize the machine's state still unless what you mean by that is to move it up one level. I've yet to find a use case where the following pattern doesn't work

I've done some additional tests and THIS IS WILD :O I totally did not realize that the suspended tree actually gets committed (with `style="display: none !important;"`)- just without component(s) that has thrown a promise. So indeed splitting the component into Outer/Inner pair works just OK. My demo that might be useful for other readers: https://codesandbox.io/s/zealous-mcnulty-9ccme?file=/src/App.js

**Reply by andria-dev** [reply]
That is interesting, I never noticed the `display: none !important` style either. I guess it makes sense since React handles each step of rendering each component so it would be able to stop midway when one component suspends and hide anything else under the `<Suspense>` component with `display: none`. Which, in turn, would allow preserving of the state of components between the suspending component and the nearest `<Suspense>`.

**Reply by Andarist** [reply]
Right - it seems that this works exactly like that but I somehow always have assumed that React only keeps the rendered output in memory and nothing actually gets committed until the whole subtree is ready. 

One slight problem with the proposed solution that I have is that it requires access to `service` but usually, we try to avoid that. I can't figure out any viable alternatives though 🤔

**Reply by andria-dev** [reply]
@Andarist I think there is a way to do it without access to `service` by using `useRef`. I'll fork your CodeSandbox and see if I can make it work and then report back here.

**Reply by andria-dev** [reply]
I did it! I almost thought it couldn't be done but it's actually kinda nice. So, what I did was create a promise but extract it's `resolve` callback to the component's function scope so I could put it into `useRef()` so I could resolve the promise from anywhere in the component at any time. I also store the promise in the `ref` so that we can throw that if necessary. Alongside these values, I placed some basic resource-style values in there — i.e. `status` and `result`.

### Getting the first value after "resolving"

Once I had those in my `ref` I was able to write an if-statement for the `condition` and if it passed I would update the `status` to "resolved" and use the `getter` to retrieve the value that the developer wants. That got us the one-time value. Now we want to subscribe to more updates.

### Updating like a subscription

In order to subscribe to future data updates, I wrote a simple `useEffect` that would check for the "resolved" status and would simply run the `getter` to provide the developer with the newest state. This would re-fire whenever the dependencies of `useEffect` would change making it act as a "subscription".

### Suspending and returning

After all of that, all that was left was to create a **Resource Reader**. For this, I'm just making an object with a read method:

```javascript
{ read() {} }
```

This handles the throwing of the promise to suspend when in the "loading" status and returning the value when in the "resolved" status.

[View the full example on CodeSandbox](https://codesandbox.io/s/xstate-suspense-prototype-0y82l?file=/src/App.js)

```javascript
function useSuspensefulGetter(state, condition, getter) {
  let resolveValue;
  const promise = new Promise((resolve) => {
    resolveValue = resolve;
  });

  const ref = useRef({
    promise,
    resolve: resolveValue,
    status: "loading",
    result: null
  });

  useEffect(() => {
    if (ref.current.status === "resolved") {
      ref.current.result = getter(state);
    }
  }, [ref, state, getter]);

  if (condition(state)) {
    ref.current.status = "resolved";
    ref.current.result = getter(state);
    ref.current.resolve();
  }

  return {
    read() {
      switch (ref.current.status) {
        case "resolved":
          return ref.current.result;
        case "loading":
        default:
          throw ref.current.promise;
      }
    }
  };
}
```
